Src/Extern/HYPRE/AMReX_HypreSolver.H

```diff
@@ -0,0 +1,665 @@
+    m_global_id.resize(m_nvars);
+    m_nrows_grid.resize(m_nvars);
+    m_id_offset.resize(m_nvars);
+    Long nrows_max = 0;
+    for (int ivar = 0; ivar < m_nvars; ++ivar) {
+        m_grids     [ivar] = amrex::convert(a_grids,m_index_type[ivar]);
+        m_local_id  [ivar].define(m_grids[ivar], m_dmap, 1, 0);
+        m_global_id [ivar].define(m_grids[ivar], m_dmap, 1, m_nghost);
+        m_nrows_grid[ivar].define(m_grids[0], m_dmap);
+        m_id_offset [ivar].define(m_grids[0], m_dmap);
+        nrows_max += m_grids[ivar].numPts();
+    }
+    m_global_id_vec.define(m_grids[0], m_dmap);
+    m_nrows.define        (m_grids[0], m_dmap);
+    AMREX_ALWAYS_ASSERT_WITH_MESSAGE(nrows_max < static_cast<Long>(std::numeric_limits<HYPRE_Int>::max()-1),
+                                     "Need to configure Hypre with --enable-bigint");
+
+    m_owner_mask.resize(m_nvars);
+    for (int ivar = 0; ivar < m_nvars; ++ivar) {
+        m_owner_mask[ivar] = amrex::OwnerMask(m_local_id[ivar], m_geom.periodicity());
+    }
+
+#ifdef AMREX_USE_GPU
+    m_cell_offset.define(m_grids[0], m_dmap);
+#endif
+
+    fill_local_id(a_marker);
+
+    // At this point, m_local_id stores the ids local to each box.
+    // m_nrows_grid stroes the number of unique points in each box.
+    // m_nrows_proc is the number of rowss for all variables on this MPI
+    // process.  If a point is invalid, its id is invalid (i.e., a very
+    // negative number).  Note that the data type of local_node_id is int,
+    // not HYPRE_Int for performance on GPU.
+
+    const int nprocs = ParallelContext::NProcsSub();
+    const int myproc = ParallelContext::MyProcSub();
+
+    Vector<HYPRE_Int> nrows_allprocs(nprocs);
+#ifdef AMREX_USE_MPI
+    if (nrows_allprocs.size() > 1) {
+        MPI_Allgather(&m_nrows_proc, sizeof(HYPRE_Int), MPI_CHAR,
+                      nrows_allprocs.data(), sizeof(HYPRE_Int), MPI_CHAR, m_comm);
+    } else
+#endif
+    {
+        nrows_allprocs[0] = m_nrows_proc;
+    }
+
+    HYPRE_Int proc_begin = 0;
+    for (int i = 0; i < myproc; ++i) {
+        proc_begin += nrows_allprocs[i];
+    }
+
+    HYPRE_Int proc_end = proc_begin;
+    for (MFIter mfi(m_nrows_grid[0]); mfi.isValid(); ++mfi) {
+        for (int ivar = 0; ivar < m_nvars; ++ivar) {
+            m_id_offset[ivar][mfi] = proc_end;
+            proc_end += m_nrows_grid[ivar][mfi];
+        }
+    }
+    AMREX_ASSERT(proc_end == proc_begin + m_nrows_proc);
+
+    fill_global_id();
+
+    // Create and initialize A, b & x
+    HYPRE_Int ilower = proc_begin;
+    HYPRE_Int iupper = proc_end-1;
+    m_hypre_ij = std::make_unique<HypreIJIface>(m_comm, ilower, iupper, m_verbose);
+    m_hypre_ij->parse_inputs(m_options_namespace);
+
+    // Obtain non-owning references to the matrix, rhs, and solution data
+    m_A = m_hypre_ij->A();
+    m_b = m_hypre_ij->b();
+    m_x = m_hypre_ij->x();
+
+    fill_matrix(a_filler);
+}
+
+template <int MSS>
+template <class Marker>
+#ifdef AMREX_USE_CUDA
```

Comment:
```suggestion
#ifdef AMREX_USE_GPU
```
to be consistent with broadening this ifdef

Reply:
It is meant for cuda compiler only.